From c35cc079994678c819aa646ed19c0b4eba17f040 Mon Sep 17 00:00:00 2001
From: GMatrixGames <gmatrixgames@hotmail.com>
Date: Thu, 11 Apr 2019 17:33:32 -0400
Subject: [PATCH] Rebrand

---
 .../command/defaults/VersionCommand.java      | 24 ++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index ee07fe5e..bc76d56c 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -214,7 +214,24 @@ public class VersionCommand extends BukkitCommand {
     private void obtainVersion() {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
-        if (version.startsWith("git-Paper-")) {
+		// Volcano start
+		if (version.startsWith("git-Volcano-")) {
+			String[] parts = version.substring("git-Volcano-".length()).split("[-\\s");
+			int distance = getDistance(null, parts[0]);
+            switch (distance) {
+                case -1:
+                    setVersionMessage("Error obtaining version information");
+                    break;
+                case 0:
+                    setVersionMessage("You are running the latest version");
+                    break;
+                case -2:
+                    setVersionMessage("Unknown version");
+                    break;
+                default:
+                    setVersionMessage("You are " + distance + " version(s) behind");
+            }
+        } else if (version.startsWith("git-Paper-")) {
             String[] parts = version.substring("git-Paper-".length()).split("[-\\s]");
             int distance = getDistance(null, parts[0]);
             switch (distance) {
@@ -271,7 +288,7 @@ public class VersionCommand extends BukkitCommand {
             return getFromJenkins(currentVer);
         } catch (NumberFormatException ex) {
             verInfo = verInfo.replace("\"", "");
-            return getFromRepo("PaperMC/Paper", "master", verInfo);
+            return getFromRepo("LavaPowered/Volcano", "master", verInfo);
         }
             /*
             BufferedReader reader = Resources.asCharSource(
@@ -293,7 +310,8 @@ public class VersionCommand extends BukkitCommand {
     private static int getFromJenkins(int currentVer) {
         try {
             BufferedReader reader = Resources.asCharSource(
-                    new URL("https://ci.destroystokyo.com/job/Paper-1.13/lastSuccessfulBuild/buildNumber"), // Paper
+                    // new URL("https://ci.destroystokyo.com/job/Paper-1.13/lastSuccessfulBuild/buildNumber"), // Paper
+					new URL("https://dev.matrixdevteam.ml/ci/job/Volcano/lastSuccessfulBuild/buildNumber"), // Volcano
                     Charsets.UTF_8
             ).openBufferedStream();
             try {
-- 
2.21.0.windows.1

