From 106fe5f2152fb0d90fd52781eb87702cbd977f06 Mon Sep 17 00:00:00 2001
From: GMatrixGames <gmatrixgames@hotmail.com>
Date: Wed, 1 May 2019 12:13:43 -0400
Subject: [PATCH] Fix versioning

---
 .../command/defaults/VersionCommand.java      | 43 ++++---------------
 1 file changed, 8 insertions(+), 35 deletions(-)

diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index d5b4078a..9ce2a06f 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -200,37 +200,14 @@ public class VersionCommand extends BukkitCommand {
             sender.sendMessage("Checking version, please wait...");
             if (!versionTaskStarted) {
                 versionTaskStarted = true;
-                new Thread(new Runnable() {
-
-                    @Override
-                    public void run() {
-                        obtainVersion();
-                    }
-                }).start();
+                new Thread(this::obtainVersion).start();
             }
         } finally {
             versionLock.unlock();
         }
     }
 
-    // TacoSpigot start = Delegate to our version-checking logic
-    private void obtainVersion() {
-        try {
-            VersionDifference difference = TacoVersioning.compareToLatest();
-            if (difference.isIdentical()) {
-                setVersionMessage("You are running the latest version");
-            } else {
-                setVersionMessage("You are" + difference + " the latest version!");
-            }
-        } catch (UnknownVersionException e) {
-            setVersionMessage("Unknown version, custom build? " + Bukkit.getVersion());
-        } catch (IOException | VersioningException e) {
-            setVersionMessage("Error obtaining version information");
-            e.printStackTrace();
-        }
-    }
 
-    /*
     // Paper start
     private void obtainVersion() {
         String version = Bukkit.getVersion();
@@ -285,8 +262,6 @@ public class VersionCommand extends BukkitCommand {
             setVersionMessage("Unknown version, custom build?");
         }
     }
-    */
-    // TacoSpigot end
 
     private void setVersionMessage(@NotNull String msg) {
         lastCheck = System.currentTimeMillis();
@@ -304,8 +279,7 @@ public class VersionCommand extends BukkitCommand {
         }
     }
 
-    // TacoSpigot start
-    /*
+
     // Paper start
     private static int getDistance(@NotNull String repo, @NotNull String verInfo) {
         try {
@@ -329,23 +303,23 @@ public class VersionCommand extends BukkitCommand {
             } finally {
                 reader.close();
             }
+             */
     }
 
     private static int getFromJenkins(int currentVer) {
         try {
-            BufferedReader reader = Resources.asCharSource(
+            // new URL("https://ci.destroystokyo.com/job/Paper-1.13/lastSuccessfulBuild/buildNumber"), // Paper
+            // Volcano
+            try (BufferedReader reader = Resources.asCharSource(
                     // new URL("https://ci.destroystokyo.com/job/Paper-1.13/lastSuccessfulBuild/buildNumber"), // Paper
-					 new URL("https://dev.matrixdevteam.ml/ci/job/Volcano/lastSuccessfulBuild/buildNumber"), // Volcano
+                    new URL("https://dev.matrixdevteam.ml/ci/job/Volcano/lastSuccessfulBuild/buildNumber"), // Volcano
                     Charsets.UTF_8
-            ).openBufferedStream();
-            try {
+            ).openBufferedStream()) {
                 int newVer = Integer.decode(reader.readLine());
                 return newVer - currentVer;
             } catch (NumberFormatException ex) {
                 ex.printStackTrace();
                 return -2;
-            } finally {
-                reader.close();
             }
         } catch (IOException e) {
             e.printStackTrace();
@@ -382,5 +356,4 @@ public class VersionCommand extends BukkitCommand {
         }
     }
     // Paper end
-    */
 }
-- 
2.21.0

